File: sources/perl/12_matrix_add/main.pl
# task 12 matrix_add — expected output: 999000000
# build: perl main.pl    run: perl main.pl
# note: flat arrays indexed i*n+j. A Perl array costs about 64 bytes an element, so the three
#       1,000,000-element arrays are roughly 64 MB each and nothing else is allocated.
use strict;
use warnings;

my $n = 1000;
my $size = $n * $n;

my (@a, @b, @c);
$#a = $size - 1;
$#b = $size - 1;
$#c = $size - 1;

for (my $i = 0; $i < $n; $i++) {
    for (my $j = 0; $j < $n; $j++) {
        my $idx = $i * $n + $j;
        $a[$idx] = $i + $j;
        $b[$idx] = $i - $j;
    }
}

for (my $i = 0; $i < $n; $i++) {
    for (my $j = 0; $j < $n; $j++) {
        my $idx = $i * $n + $j;
        $c[$idx] = $a[$idx] + $b[$idx];
    }
}

my $total = 0;
for (my $i = 0; $i < $n; $i++) {
    for (my $j = 0; $j < $n; $j++) {
        $total += $c[$i * $n + $j];
    }
}

print $total, "\n";

File: sources/java/12_matrix_add/Main.java
// task 12 matrix_add — expected output: 999000000
// build: javac Main.java    run: java Main
// build (graalvm native-image): native-image -O2 Main    run: ./Main
// Flat long[n*n] arrays with index i*n+j.

public class Main {
    public static void main(String[] args) {
        final int n = 1000;
        long[] a = new long[n * n];
        long[] b = new long[n * n];
        long[] c = new long[n * n];

        for (int i = 0; i < n; i++) {
            int row = i * n;
            for (int j = 0; j < n; j++) {
                a[row + j] = i + j;
                b[row + j] = i - j;
            }
        }
        for (int i = 0; i < n; i++) {
            int row = i * n;
            for (int j = 0; j < n; j++) {
                c[row + j] = a[row + j] + b[row + j];
            }
        }

        long total = 0;
        for (int idx = 0; idx < n * n; idx++) {
            total += c[idx];
        }
        System.out.println(total);
    }
}
